# UMAP: Uniform Manifold Approximation and Projection
## Inteligencia Computacional 2021-2, Grupo 8a
Nicolás Canales, Matías Vergara

Este notebook tiene por objetivo aplicar UMAP sobre las curvas de luz con características computadas para reducir su dimensionalidad a 2D, y luego visualizar el resultado mediante un scatterplot en búsqueda de clusters.

Recordemos que los objetos con los que estamos trabajando son aquellos de tipo periódico, clasificados por ALeRCE como: "LPV", "Periodic-Other", "RRL", "CEP", "E" o "DSCT". 

### Referencias:
umap-learn 0.5.1 Project description on pypi: https://pypi.org/project/umap-learn/

UMAP learn the docs - How to use UMAP https://umap-learn.readthedocs.io/en/latest/basic_usage.html


## Instalación de librerías necesarias

In [3]:
!pip install umap-learn[plot]
!pip install pandas_profiling
!pip install jupyter
!pip install iprogress
!pip install summarytools

     |████████████████████████████████| 80 kB 2.4 MB/s 
     |████████████████████████████████| 1.1 MB 7.0 MB/s 
     |████████████████████████████████| 15.8 MB 646 bytes/s 
     |████████████████████████████████| 76 kB 5.1 MB/s 
     |████████████████████████████████| 791 kB 34.7 MB/s 
     |████████████████████████████████| 125 kB 52.1 MB/s 
     |████████████████████████████████| 786 kB 41.0 MB/s 
     |████████████████████████████████| 779 kB 30.8 MB/s 
     |████████████████████████████████| 778 kB 45.2 MB/s 
     |████████████████████████████████| 776 kB 44.4 MB/s 
     |████████████████████████████████| 769 kB 38.2 MB/s 
     |████████████████████████████████| 766 kB 45.6 MB/s 
     |████████████████████████████████| 1.0 MB 37.1 MB/s 
     |████████████████████████████████| 722 kB 42.3 MB/s 
     |████████████████████████████████| 722 kB 43.7 MB/s 
     |████████████████████████████████| 715 kB 52.6 MB/s 
     |████████████████████████████████| 705 kB 48.9 MB/s 
     |██████████

In [4]:
import numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
%matplotlib inline
import umap 
from summarytools.summarytools import dfSummary
import pandas_profiling
import re
from io import BytesIO
from PIL import Image
import base64
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import HoverTool, ColumnDataSource, CategoricalColorMapper
from bokeh.palettes import Spectral10, Spectral6


## Traer las curvas de luz con features

In [ ]:
# traemos las curvas de luz con sus features
!gdown --id 19uB-u0gYCGKFlKFXCKIwsK5G4_MCFvV1
# traemos además las curvas de luz sintéticas, que deberemos unir a nuestras curvas
# originales
!gdown --id 1r8BcRI5vJgi5s9qOGsja_mlAk3S6fW7P
# traemos tambien el archivo labels.csv, con las clasificaciones
!gdown --id  1LU1sCIVXO8BQRMeKCCqu1vZGnceV6P5c

Downloading...
From: https://drive.google.com/uc?id=19uB-u0gYCGKFlKFXCKIwsK5G4_MCFvV1
To: /content/lc_features.csv
100% 143M/143M [00:00<00:00, 168MB/s]
Downloading...
From: https://drive.google.com/uc?id=1r8BcRI5vJgi5s9qOGsja_mlAk3S6fW7P
To: /content/sintetic_features.csv
100% 4.23M/4.23M [00:00<00:00, 136MB/s]
Downloading...
From: https://drive.google.com/uc?id=1LU1sCIVXO8BQRMeKCCqu1vZGnceV6P5c
To: /content/labels_set.csv
100% 10.2M/10.2M [00:00<00:00, 90.1MB/s]


## Preparando la data para UMAP

In [ ]:
lc_features = pd.read_csv("lc_features.csv", sep=',', index_col=1)
sintetic_features = pd.read_csv("sintetic_features.csv", index_col=1)
labels = pd.read_csv("labels_set.csv", index_col=0)

print(len(lc_features.index))
print(len(sintetic_features.index))

87003
2606


Se logró calcular features para 87.003 curvas de luz originales y 2606 sintéticas, sin embargo, muchas de ellas incorporan valores NaN. Si realizamos un `dropna()`, quedaremos con poco más de 20.000! (que es perder demasiada información). Necesitamos encontrar otra forma de tratar este problema.

Un camino viable puede ser ver qué características son no-nulas para la mayoría de los datos. Para ello usamos pandas-profiler para generar un análisis de las variables, mediante la siguiente celda de código (comentada).

Importante: pandas profiling requiere la versión 0.25 de pandas, que no es la que trae Colab por defecto. Se requiere reiniciar el entorno una vez instalada.

In [ ]:
#!pip install pandas==0.25
#profile = pandas_profiling.ProfileReport(lc_features)
#profile.to_file("profile_lc_features.html")

Lo anterior nos genera el archivo `profile_lc_features.html`, el cual estudiamos para dar solución al problema anterior. En particular podemos notar que:
- Eta_e_g presenta un 48.8% de missing values
- MaxSlope_g presenta un 48.8% de missing values
- Eta_e_r presenta un 52.3% de missing values
- MaxSlope_r presenta un 52.3% de missing values

Probaremos dropeando estas columnas antes de aplicar el `drop.na()`, para ver  si esto reduce la fracción de data perdida.

In [ ]:
lc_features.drop(axis=1, labels=['Eta_e_g', 'Eta_e_r', 'MaxSlope_g', 'MaxSlope_r'], inplace=True)
sintetic_features.drop(axis=1, labels=['Eta_e_g', 'Eta_e_r', 'MaxSlope_g', 'MaxSlope_r'], inplace=True)

Veamos ahora si cuántos datos mantenemos al aplicar `drop.na()`

---



In [ ]:
lc_features.dropna(inplace=True)
sintetic_features.dropna(inplace=True)
print(len(lc_features.index))
print(len(sintetic_features.index))

74908
2173


Retenemos 77081 filas! Es decir, solo perdimos alrededor del 13.1% de la data, lo cual es bueno. Sin embargo, también debemos ver que los datos que perdimos no correspondan a una sola clase (o aún peor: a nuestras clases menos representadas). Veamos cuantas entradas tenemos por clase:

In [ ]:
count = {"LPV":0, "Periodic-Other": 0, "RRL": 0, "CEP": 0, "E": 0, "DSCT": 0}
for oid in lc_features.index.unique():
  try:
    alerce_class = labels.loc[oid].values[0]
    count[alerce_class] += 1
  except:
    pass
for oid in sintetic_features.index.unique():
    alerce_class = oid[8:]
    alerce_class = "".join(re.findall("[a-zA-Z]+", alerce_class))
    if alerce_class == "PeriodicOther":
      alerce_class = "Periodic-Other"
    elif alerce_class == "DSCTS":
      alerce_class = "DSCT"
    else:
      alerce_class = "CEP"
    count[alerce_class] += 1
count


{'CEP': 906,
 'DSCT': 1336,
 'E': 33486,
 'LPV': 9799,
 'Periodic-Other': 2104,
 'RRL': 29450}

Recordemos que, anteriormente, la distribución era la siguiente (Véase notebook de data-augmentation):

{'CEP': 1236,
 'DSCT': 1464,
 'E': 37900,
 'LPV': 14045,
 'Periodic-Other': 2512,
 'RRL': 32464}
 
 Notamos que hemos perdido principalmente entradas de RRL y LPV, dos de las clases más representadas. Asumiremos que podemos continuar sin problema.


Hasta ahora nuestra data es de la forma `oid, feature1, feature2, ..., featuren`. Esto es suficiente para que UMAP realice la reducción de dimensionalidad y nos permita visualizar los datos... Pero hay un problema: no tenemos incorporada la clase de nuestros datos, el **target!**.


**IMPORTANTE: Deseamos incluir el target en los datos para poder colorear la visualización que resultará de aplicar UMAP, ¡no para entrenar!**


Para dar solución a esta situación volveremos a generar un cruce de datos, esta vez entre los datasets `labels_set.csv` y nuestro dataframe de curvas de luz originales y sintéticas, rescatando la clase en ALeRCE `classALeRCE` de cada curva para incorporarla en el dataframe y luego usar dicha columna como target de color en la visualización que obtengamos mediante UMAP. Para las curvas sintéticas - que no están incluidas en el labels_set - obtendremos su clase a partir de su nombre.

In [ ]:
# Primero agregamos target a las curvas originales
lc_features["target"] = ""
for index, row in lc_features.iterrows():
    alerce_class = labels.loc[index].values[0]
    lc_features['target'][index] = alerce_class

# Y lo mismo para las curvas sintéticas
sintetic_features["target"] = ""
for index, row in sintetic_features.iterrows():
    alerce_class = index[8:]
    alerce_class = "".join(re.findall("[a-zA-Z]+", alerce_class))
    if alerce_class == "PeriodicOther":
      alerce_class = "Periodic-Other"
    elif alerce_class == "DSCTS":
      alerce_class = "DSCT"
    else:
      alerce_class = "CEP"
    sintetic_features['target'][index] = alerce_class

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [ ]:
lc_features.head()

Unnamed: 0  Multiband_period  ...  LinearTrend_r  target
oid                                         ...                       
ZTF17aadmidf           0          0.149331  ...       0.000069       E
ZTF17aadqpnu           0          0.544647  ...       0.000121     RRL
ZTF18aapkwuy           0          0.578986  ...      -0.000026     RRL
ZTF18aajoeri           0          0.620327  ...      -0.000089     RRL
ZTF18adaiqlm           0          0.599899  ...      -0.000296       E

[5 rows x 100 columns]

In [ ]:
sintetic_features.tail()

Unnamed: 0  ...          target
oid                                   ...                
sinteticPeriodicOther633           0  ...  Periodic-Other
sinteticDSCTS491                   0  ...            DSCT
sinteticDSCTS8                     0  ...            DSCT
sinteticDSCTS92                    0  ...            DSCT
sinteticDSCTS485                   0  ...            DSCT

[5 rows x 100 columns]

Ahora que ya tenemos ambos dataset con sus respectivos target y las mismas columnas, podemos proceder a concatenarlos y entrar a trabajar con UMAP.

In [ ]:
data = pd.concat([lc_features, sintetic_features])
print(len(data.index))
data.to_csv("augmented_features.csv")

77081


## UMAP en acción

In [ ]:
# Reproducibilidad: en caso de querer ejecutar solamente esta sección correr esta celda
# para descargar la data concatenada en la sección anterior.
!gdown --id 1HFEbip5SX591MCLi-S6DKw7LEx-CJFNt

Downloading...
From: https://drive.google.com/uc?id=1HFEbip5SX591MCLi-S6DKw7LEx-CJFNt
To: /content/augmented_features.csv
100% 131M/131M [00:00<00:00, 144MB/s]


Para usar UMAP para esta tarea primero necesitamos construir un objeto UMAP que hará el trabajo por nosotros. Para eso basta instanciar la clase:

In [ ]:
reducer = umap.UMAP()


Antes de hacer cualquier trabajo con los datos será útil limpiarlos un poco. Dado que las medidas están en escalas completamente distintas, será útil convertir cada feature en z-scores (cantidad de desviaciones desde la media) para poder compararlas.

Desde luego, antes de cualquier procesamiento también nos preocupamos de dropear la columna de `target`, pues buscamos generar un modelo no supervisado.

In [ ]:
data = pd.read_csv("augmented_features.csv", index_col=0)
scaled_data = StandardScaler().fit_transform(data.drop(labels='target', axis=1))

Ahora necesitamos entrenar nuestro reductor, permitiéndole aprender del manifold. Para ello UMAP sigue la API de sklearn, incorporando el método fit al cual le pasamos la data de la cual queremos que el modelo aprenda.


In [ ]:
reducer = umap.UMAP(random_state=13)
reducer.fit(scaled_data)

UMAP(a=None, angular_rp_forest=False, b=None, dens_frac=0.0, dens_lambda=0.0,
     dens_var_shift=0.1, densmap=False, disconnection_distance=None,
     force_approximation_algorithm=False, init='spectral', learning_rate=1.0,
     local_connectivity=1.0, low_memory=True, metric='euclidean',
     metric_kwds=None, min_dist=0.1, n_components=2, n_epochs=None, n_jobs=-1,
     n_neighbors=15, negative_sample_rate=5, output_dens=False,
     output_metric='euclidean', output_metric_kwds=None, random_state=13,
     repulsion_strength=1.0, set_op_mix_ratio=1.0, spread=1.0,
     target_metric='categorical', target_metric_kwds=None, ...)

Luego tenemos el método transform, que nos dará la data transformada.

In [ ]:
embedding = reducer.transform(scaled_data)

Finalmente, importamos algunas librerías de visualización y armamos el scatterplot de nuestros datos.

In [ ]:
output_notebook()

In [ ]:
data_df = pd.DataFrame(embedding, columns=('x', 'y'))
data_df['target'] = [x for x in data.target]

datasource = ColumnDataSource(data_df)
color_mapping = CategoricalColorMapper(factors=["E", "RRL", "CEP", "DSCT", "LPV", "Periodic-Other"],
                                       palette=Spectral6)

plot_figure = figure(
    title='UMAP projection of the periodic light curves',
    plot_width=600,
    plot_height=600,
    tools=('pan, wheel_zoom, reset')
)

plot_figure.circle(
    'x',
    'y',
    source=datasource,
    color=dict(field='target', transform=color_mapping),
    line_alpha=0.6,
    fill_alpha=0.6,
    size=4,
    legend='target'
)
show(plot_figure)

In [ ]:
embedding.tofile("embedding_UMAP.csv")

## Y si antes de UMAP, escogemos las features más importantes...?

Siguiendo la instrucción del enunciado del proyecto, probaremos ahora qué pasa si ayudamos al modelo trabajando solo con un subconjunto de las features. 

Para escoger las features, se estudia la distribución, pérdida de entradas y correlación entre columnas mediante  la generación de un `pandas profiler` de los datos.

In [76]:
# Reproducibilidad: en caso de querer ejecutar solamente esta sección correr esta celda
# para descargar la data concatenada en la sección 1.
!gdown --id 1HFEbip5SX591MCLi-S6DKw7LEx-CJFNt #augmented_features, curvas de luz con sus features
                                              #incluye curvas originales y sintéticas
!gdown --id 1XCl8BiVOP7aheBYjOHIAM378s_34-8kl #reduced_data, misma data pero subsampleando las clases
                                              #sobrerepresentadas

Downloading...
From: https://drive.google.com/uc?id=1HFEbip5SX591MCLi-S6DKw7LEx-CJFNt
To: /content/augmented_features.csv
100% 131M/131M [00:01<00:00, 93.1MB/s]
Downloading...
From: https://drive.google.com/uc?id=1XCl8BiVOP7aheBYjOHIAM378s_34-8kl
To: /content/reduced_data.csv
100% 23.9M/23.9M [00:00<00:00, 111MB/s]


In [102]:
data = pd.read_csv("augmented_features.csv", index_col=0)

Generamos el profiler

In [78]:
#!pip install pandas==0.25
#el profiler solo es compatible con pandas 0.25
#se recomienda obtenerlo y luego reiniciar el entorno
#profile = pandas_profiling.ProfileReport(data)
#profile.to_file("profile_data_augmented.html")

# o para traer el resultado desde el drive del proyecto:
!gdown --id 12IZp0_0A6yKDQdGfIQUe3F4F8ZlclItB

Downloading...
From: https://drive.google.com/uc?id=12IZp0_0A6yKDQdGfIQUe3F4F8ZlclItB
To: /content/profile_data_augmented.html
100% 3.49M/3.49M [00:00<00:00, 111MB/s]


En la sección "Overview" del profiler generado, se presentan las caraceristicas que el profiler propone descartar, ya sea porque tienen un alto porcentaje de ceros o missing values, porque son demasiado sesgadas (skew) y porque tienen mucha correlación. Se realizan pruebas con distintos subconjuntos de features, teniendo en cuenta estas recomendaciones del profiler.

In [111]:
interest_features = [
                     'Multiband_period',
                     'delta_period_g',
                     'delta_period_r',
                 #    'GP_DRW_sigma_r', demasiado skew
                 #    'GP_DRW_tau_g',
                 #    'GP_DRW_sigma_r',
                 #    'GP_DRW_tau_r',
                     'Harmonics_mag_1_g',
                     'Harmonics_mag_1_r',
                     'Harmonics_mse_r', # comentar esta da otro conjunto viable
                     'Harmonics_mse_g', ##
                     #'Power_rate_1/4', ##
                     #'AndersonDarling_g', ##
                     #'AndersonDarling_g' ##
                     'iqr_g',
                     'iqr_r',
                     'Amplitude_g',
                     'Mean_g',
                     'Meanvariance_g',
                     'Amplitude_r',
                     'Mean_r',
                     'Meanvariance_r',
                     'PairSlopeTrend_r',
                     'target',
                     'LinearTrend_r',
                     'ExcessVar_r',
                     'LinearTrend_g',
                     'ExcessVar_g'
                     
]


data = data[interest_features]
data.head()

Multiband_period  delta_period_g  ...  LinearTrend_g  ExcessVar_g
oid                                             ...                            
ZTF17aadmidf          0.149331        0.000000  ...      -0.000188     0.000050
ZTF17aadqpnu          0.544647        0.000000  ...       0.000037     0.000400
ZTF18aapkwuy          0.578986        0.000000  ...       0.000070     0.000437
ZTF18aajoeri          0.620327        0.000000  ...      -0.000095     0.000248
ZTF18adaiqlm          0.599899        0.369112  ...       0.001207     0.000074

[5 rows x 21 columns]

### Repetimos el proceso con nuestra data de features reducidas manualmente.

In [112]:
reducer = umap.UMAP()
scaled_data = StandardScaler().fit_transform(data.drop(labels='target', axis=1))
reducer = umap.UMAP(random_state=13)

In [113]:
reducer.fit(scaled_data)
embedding = reducer.transform(scaled_data)

Visualicemos los resultados de UMAP en este nuevo intento

In [109]:
output_notebook()

In [114]:
data_df = pd.DataFrame(embedding, columns=('x', 'y'))
data_df['target'] = [x for x in data.target]

datasource = ColumnDataSource(data_df)
color_mapping = CategoricalColorMapper(factors=["E", "RRL", "CEP", "DSCT", "LPV", "Periodic-Other"],
                                       palette=Spectral6)

plot_figure = figure(
    title='UMAP projection of the periodic light curves',
    plot_width=600,
    plot_height=600,
    tools=('pan, wheel_zoom, reset')
)

plot_figure.circle(
    'x',
    'y',
    source=datasource,
    color=dict(field='target', transform=color_mapping),
    line_alpha=0.6,
    fill_alpha=0.6,
    size=4,
    legend='target'
)
show(plot_figure)